lua/lsp/null-ls.lua

```diff
@@ -1,141 +1,136 @@
-      -- FIXME: why doesn't this work?
-      -- builtin_formatter._opts.args = formatter.args or builtin_formatter._opts.args
-      -- builtin_formatter._opts.to_stdin = formatter.stdin or builtin_formatter._opts.to_stdin
-      local resolved_path = validate_provider_request(builtin_formatter)
-      if resolved_path then
-        builtin_formatter._opts.command = resolved_path
-        table.insert(M.requested_providers, builtin_formatter)
-        Log:get_default().info("Using format provider", builtin_formatter.name)
+    if not builtin_formatter then
+      logger.error("Not a valid formatter:", formatter.exe)
+    elseif not vim.tbl_contains(M.langs[filetype].registered.formatters, builtin_formatter) then
+      local formatter_cmd = find_command(builtin_formatter._opts.command)
+      if not formatter_cmd then
+        logger.warn("Not found:", builtin_formatter._opts.command)
+        table.insert(M.langs[filetype].errors.formatters, formatter.exe)
       else
-        -- mark it here to avoid re-doing the lookup again
-        register_failed_request(filetype, formatter.exe, "formatters")
+        logger.info("Using formatter:", formatter_cmd)
+        table.insert(M.langs[filetype].registered.formatters, builtin_formatter.with { command = formatter_cmd })
       end
     end
   end
 
   for _, linter in pairs(lvim.lang[filetype].linters) do
-    local builtin_diagnoser = null_ls.builtins.diagnostics[linter.exe]
-    Log:get_default().debug("validating lint provider: ", linter.exe)
-    -- special case: fallback to "eslint"
-    -- https://github.com/jose-elias-alvarez/null-ls.nvim/blob/9b8458bd1648e84169a7e8638091ba15c2f20fc0/doc/BUILTINS.md#eslint
-    -- if provider.exe
-    if linter.exe == "eslint_d" then
-      builtin_diagnoser = null_ls.builtins.diagnostics.eslint.with { command = "eslint_d" }
-    end
-    if not vim.tbl_contains(M.requested_providers, builtin_diagnoser) then
-      -- FIXME: why doesn't this work?
-      -- builtin_diagnoser._opts.args = linter.args or builtin_diagnoser._opts.args
-      -- builtin_diagnoser._opts.to_stdin = linter.stdin or builtin_diagnoser._opts.to_stdin
-      local resolved_path = validate_provider_request(builtin_diagnoser)
-      if resolved_path then
-        builtin_diagnoser._opts.command = resolved_path
-        table.insert(M.requested_providers, builtin_diagnoser)
-        Log:get_default().info("Using linter provider", builtin_diagnoser.name)
+    local builtin_linter = adapt_linter(linter)
+    if not builtin_linter then
+      logger.error("Not a valid linter:", linter.exe)
+    elseif not vim.tbl_contains(M.langs[filetype].registered.linters, builtin_linter) then
+      local linter_cmd = find_command(builtin_linter._opts.command)
+      if not linter_cmd then
+        logger.warn("Not found:", builtin_linter._opts.command)
+        table.insert(M.langs[filetype].errors.linters, linter.exe)
       else
```

Comment:
no need for the else part if you're doing `if not ..`

Reply:
If the linter is not found, then add it as failed, else add it as registered.